--- modules/http-proxy.c	2007-08-27 13:05:18.000000000 +0300
+++ modules/http-proxy.c	2007-08-27 12:55:28.000000000 +0300
@@ -52,7 +52,7 @@
 
 /* Global variables used by the HTTP proxy config */
 static GConfClient *gl_client = NULL;
-static GMutex *gl_mutex = NULL;	/* This mutex protects preference values
+G_LOCK_DEFINE_STATIC(gl_mutex);	/* This mutex protects preference values
 				 * and ensures serialization of authentication
 				 * hook callbacks       
 				 */
@@ -247,7 +247,7 @@
 
     ret = proxy_should_for_hostname (toplevel_uri->host_name);
 
-    g_mutex_lock (gl_mutex);
+    G_LOCK(gl_mutex);
 
     if (ret && gl_http_proxy != NULL) {
 	ret = host_port_from_string(gl_http_proxy, &(proxy_info->host),
@@ -263,7 +263,7 @@
 	ret = FALSE;
     }
 
-    g_mutex_unlock(gl_mutex);
+    G_UNLOCK(gl_mutex);
 
     return ret;
 }
@@ -499,7 +499,7 @@
 	|| strcmp(key, KEY_GCONF_HTTP_PROXY_PORT) == 0) {
 	gboolean use_proxy_value;
 
-	g_mutex_lock(gl_mutex);
+	G_LOCK(gl_mutex);
 
 	/* Check and see if we are using the proxy */
 	use_proxy_value =
@@ -507,20 +507,20 @@
 				  KEY_GCONF_USE_HTTP_PROXY, NULL);
 	construct_gl_http_proxy(use_proxy_value);
 
-	g_mutex_unlock(gl_mutex);
+	G_UNLOCK(gl_mutex);
     } else if (strcmp(key, KEY_GCONF_HTTP_AUTH_USER) == 0
 	       || strcmp(key, KEY_GCONF_HTTP_AUTH_PW) == 0
 	       || strcmp(key, KEY_GCONF_HTTP_USE_AUTH) == 0) {
 	gboolean use_proxy_auth;
 
-	g_mutex_lock(gl_mutex);
+	G_LOCK(gl_mutex);
 
 	use_proxy_auth =
 	    gconf_client_get_bool(gl_client,
 				  KEY_GCONF_HTTP_USE_AUTH, NULL);
 	set_proxy_auth(use_proxy_auth);
 
-	g_mutex_unlock(gl_mutex);
+	G_UNLOCK(gl_mutex);
     }
 }
 
@@ -532,7 +532,6 @@
     gboolean use_proxy_auth;
 
     gl_client = gconf_client_get_default();
-    gl_mutex = g_mutex_new();
 
     gconf_client_add_dir(gl_client, PATH_GCONF_GNOME_VFS,
 			 GCONF_CLIENT_PRELOAD_ONELEVEL, &gconf_error);

