Index: libgnomevfs/xdgmimeglob.c
===================================================================
RCS file: /cvs/gnome/gnome-vfs/libgnomevfs/xdgmimeglob.c,v
retrieving revision 1.11
diff -u -p -w -r1.11 xdgmimeglob.c
--- libgnomevfs/xdgmimeglob.c	1 Dec 2005 19:32:02 -0000	1.11
+++ libgnomevfs/xdgmimeglob.c	21 Feb 2006 19:42:37 -0000
@@ -263,7 +263,7 @@ _xdg_glob_hash_insert_text (XdgGlobHashN
 		{
 		  child = _xdg_glob_hash_node_new ();
 		  child->character = '\000';
-		  child->mime_type = mime_type;
+		  child->mime_type = strdup (mime_type);
 		  child->child = NULL;
 		  child->next = node->child;
 		  node->child = child;
@@ -272,7 +272,7 @@ _xdg_glob_hash_insert_text (XdgGlobHashN
 	}
       else
 	{
-	  node->mime_type = mime_type;
+	  node->mime_type = strdup (mime_type);
 	}
     }
   else
@@ -475,7 +475,7 @@ _xdg_glob_hash_append_glob (XdgGlobHash 
       glob_hash->literal_list = _xdg_glob_list_append (glob_hash->literal_list, strdup (glob), strdup (mime_type));
       break;
     case XDG_GLOB_SIMPLE:
-      glob_hash->simple_node = _xdg_glob_hash_insert_text (glob_hash->simple_node, glob + 1, strdup (mime_type));
+      glob_hash->simple_node = _xdg_glob_hash_insert_text (glob_hash->simple_node, glob + 1, mime_type);
       break;
     case XDG_GLOB_FULL:
       glob_hash->full_list = _xdg_glob_list_append (glob_hash->full_list, strdup (glob), strdup (mime_type));
